plancheck/expect_resource_action_test.go

```diff
@@ -0,0 +1,531 @@
+package plancheck_test
+
+import (
+	"regexp"
+	"testing"
+
+	r "github.com/hashicorp/terraform-plugin-testing/helper/resource"
+	"github.com/hashicorp/terraform-plugin-testing/plancheck"
+)
+
+func Test_ExpectedResourceAction_NoOp(t *testing.T) {
+	t.Parallel()
+
+	r.Test(t, r.TestCase{
+		ExternalProviders: map[string]r.ExternalProvider{
+			"random": {
+				Source: "registry.terraform.io/hashicorp/random",
+			},
+		},
+		Steps: []r.TestStep{
+			{
+				Config: `resource "random_string" "one" {
+					length = 16
+				}`,
+			},
+			{
+				Config: `resource "random_string" "one" {
+					length = 16
+				}`,
+				ConfigPlanChecks: r.ConfigPlanChecks{
+					PreApply: []plancheck.PlanCheck{
+						plancheck.ExpectResourceAction("random_string.one", plancheck.ResourceActionNoop),
+					},
+				},
+			},
+		},
+	})
+}
+
+func Test_ExpectedResourceAction_NoOp_NoMatch(t *testing.T) {
+	t.Parallel()
+
+	r.Test(t, r.TestCase{
+		ExternalProviders: map[string]r.ExternalProvider{
+			"random": {
+				Source: "registry.terraform.io/hashicorp/random",
+			},
+		},
+		Steps: []r.TestStep{
+			{
+				Config: `resource "random_string" "one" {
+					length = 16
+				}`,
+				ConfigPlanChecks: r.ConfigPlanChecks{
+					PreApply: []plancheck.PlanCheck{
+						plancheck.ExpectResourceAction("random_string.one", plancheck.ResourceActionNoop),
+					},
+				},
+				ExpectError: regexp.MustCompile(`expected NoOp, got action\(s\): \[create\]`),
+			},
+		},
+	})
+}
+
+func Test_ExpectedResourceAction_Create(t *testing.T) {
+	t.Parallel()
+
+	r.Test(t, r.TestCase{
+		ExternalProviders: map[string]r.ExternalProvider{
+			"random": {
+				Source: "registry.terraform.io/hashicorp/random",
+			},
+		},
+		Steps: []r.TestStep{
+			{
+				Config: `resource "random_string" "one" {
+					length = 16
+				}`,
+				ConfigPlanChecks: r.ConfigPlanChecks{
+					PreApply: []plancheck.PlanCheck{
+						plancheck.ExpectResourceAction("random_string.one", plancheck.ResourceActionCreate),
+					},
+				},
+			},
+		},
+	})
+}
+
+func Test_ExpectedResourceAction_Create_NoMatch(t *testing.T) {
+	t.Parallel()
+
+	r.Test(t, r.TestCase{
+		ExternalProviders: map[string]r.ExternalProvider{
+			"random": {
+				Source: "registry.terraform.io/hashicorp/random",
+			},
+		},
+		Steps: []r.TestStep{
+			{
+				Config: `resource "random_string" "one" {
+					length = 16
+				}`,
+			},
+			{
+				Config: `resource "random_string" "one" {
+					length = 15
+				}`,
+				ConfigPlanChecks: r.ConfigPlanChecks{
+					PreApply: []plancheck.PlanCheck{
+						plancheck.ExpectResourceAction("random_string.one", plancheck.ResourceActionCreate),
+					},
+				},
+				ExpectError: regexp.MustCompile(`expected Create, got action\(s\): \[delete create\]`),
+			},
+		},
+	})
+}
+
+func Test_ExpectedResourceAction_Read(t *testing.T) {
```

Comment:
I don't think it hurts to have another test, although I'm not 100% sure how to trigger a `Read` action for a resource. This test only generates a `Read` action for the data source because it needs to be read during the apply since it's unknown.

Reply:
That would explain my troubles generating this test. Thanks for the context Brian!

Reply:
Thanks for the clarification.